### PR TITLE
Add nested categories for fit functions

### LINKED ISSUE #27586: Add nested categories to fit functions
Can you reuse the algorithms selection widgets?

### PR DESCRIPTION
**Pull request by alicerussell1**
**Description of work.**
This PR adds the functionality for fit functions to be stored in nested categories, using the separator `\\` (the same as algorithms). In making these changes it also means the functions can appear in multiple categories (uses `;` as a separator)

**To test:**
Change the category of an existing function to be a nested category e.g. `General\\NewCategory`
Open Workbench
Load a workspace and do a plot.
Click fit and right click on the propertybrowser and click `add function`
The function you changed should appear in a nested category.

Fixes #27586 

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: qt/widgets/common/src/SelectFunctionDialog.cpp [hunk 1]
```diff
@@ -89,14 +89,62 @@ void SelectFunctionDialog::constructFunctionTree(
...
+        QTreeWidgetItem *catItem = nullptr;
+        int n = subCategories.size();
+        bool show = false;
+        for (int j = 0; j < n; j++) {
+          if (showCategory(subCategories[j].toStdString())) {
+            show = true;
+          }
+        }
+        if (show) {
+          for (int j = 0; j < n; j++) {
```
Comment (StephenSmith25) [inline qt/widgets/common/src/SelectFunctionDialog.cpp]: Same as above.

File: qt/widgets/common/src/SelectFunctionDialog.cpp [hunk 2]
```diff
@@ -89,14 +89,62 @@ void SelectFunctionDialog::constructFunctionTree(
...
+            fit->setText(0, QString::fromStdString(function));
+          }
+        }
+      } else {
+        // go through the path and add the folders if they don't already exist
+        QString currentPath = subCategories[0];
+        QTreeWidgetItem *catItem = nullptr;
+        int n = subCategories.size();
+        bool show = false;
+        for (int j = 0; j < n; j++) {
```
Comment (StephenSmith25) [inline qt/widgets/common/src/SelectFunctionDialog.cpp]: Mostly inconsequential, but the preference in the standards is to favor preincrement-ing, so

`for (int j = 0 ; j < n ; ++j)`

File: qt/widgets/common/src/SelectFunctionDialog.cpp [hunk 3]
```diff
@@ -89,14 +89,62 @@ void SelectFunctionDialog::constructFunctionTree(
...
+          for (const auto &function : entry.second) {
+            QTreeWidgetItem *fit = new QTreeWidgetItem(catItem);
+            fit->setText(0, QString::fromStdString(function));
+          }
+        }
+      } else {
+        // go through the path and add the folders if they don't already exist
+        QString currentPath = subCategories[0];
+        QTreeWidgetItem *catItem = nullptr;
+        int n = subCategories.size();
```
Comment (StephenSmith25) [inline qt/widgets/common/src/SelectFunctionDialog.cpp]: Would help for readability if something more descriptive was used in place of `int n`, especially as its used further down in the block.

**Review by StephenSmith25 — CHANGES_REQUESTED** [review]
Changes work as described, good work. Just a few minor comments

File: qt/widgets/common/src/SelectFunctionDialog.cpp (hunk 1 above)
Comment (alicerussell1) [reply qt/widgets/common/src/SelectFunctionDialog.cpp]: Ah yes I've fallen into old habits

**Review by StephenSmith25 — APPROVED** [review]
Looks good to go now.

**Pull request merged**